Integrate with givPower balance aggregator microservice

mohammadranjbarz opened the pull request:
related to #1094

dependabot[bot] committed:
Bump vm2 from 3.9.17 to 3.9.19

Bumps [vm2](https://github.com/patriksimek/vm2) from 3.9.17 to 3.9.19.
- [Release notes](https://github.com/patriksimek/vm2/releases)
- [Changelog](https://github.com/patriksimek/vm2/blob/master/CHANGELOG.md)
- [Commits](https://github.com/patriksimek/vm2/compare/3.9.17...3.9.19)

---
updated-dependencies:
- dependency-name: vm2
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump @apollo/server from 4.3.1 to 4.7.4

Bumps [@apollo/server](https://github.com/apollographql/apollo-server/tree/HEAD/packages/server) from 4.3.1 to 4.7.4.
- [Release notes](https://github.com/apollographql/apollo-server/releases)
- [Changelog](https://github.com/apollographql/apollo-server/blob/main/packages/server/CHANGELOG.md)
- [Commits](https://github.com/apollographql/apollo-server/commits/@apollo/server@4.7.4/packages/server)

---
updated-dependencies:
- dependency-name: "@apollo/server"
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump semver from 5.7.1 to 5.7.2

Bumps [semver](https://github.com/npm/node-semver) from 5.7.1 to 5.7.2.
- [Release notes](https://github.com/npm/node-semver/releases)
- [Changelog](https://github.com/npm/node-semver/blob/v5.7.2/CHANGELOG.md)
- [Commits](https://github.com/npm/node-semver/compare/v5.7.1...v5.7.2)

---
updated-dependencies:
- dependency-name: semver
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump word-wrap from 1.2.3 to 1.2.4

Bumps [word-wrap](https://github.com/jonschlinkert/word-wrap) from 1.2.3 to 1.2.4.
- [Release notes](https://github.com/jonschlinkert/word-wrap/releases)
- [Commits](https://github.com/jonschlinkert/word-wrap/compare/1.2.3...1.2.4)

---
updated-dependencies:
- dependency-name: word-wrap
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

CarlosQ96 committed:
optimize latest project update query

mohammadranjbarz committed:
Integrate with givPower balance aggregator microservice

related to #1094

mohammadranjbarz committed:
Integrate with givPower balance aggregator microservice

related to #1094

mohammadranjbarz committed:
Fix failed test cases

related to #1094

mohammadranjbarz committed:
Fix failed test cases

related to #1094

mohammadranjbarz committed:
WIP test cases dont work

related to #1094

mohammadranjbarz committed:
WIP test cases dont work

related to #1094

mohammadranjbarz committed:
WIP test cases dont work

related to #1094

mohammadranjbarz committed:
WIP test cases dont work

related to #1094

mohammadranjbarz committed:
Merge branch 'staging' into f_1094_integrate_with_balance_aggregator_microservice

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+  date: Date; // sample: "2023-08-10T16:18:02.655Z"
+}
+
+export interface GivPowerBalanceAggregatorInterface {
```
I would name it
```suggestion
export interface IGivPowerBalanceAggregator {
```

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorMock.ts:
```diff
@@ -0,0 +1,60 @@
+  GivPowerBalanceAggregatorInterface,
+} from './givPowerBalanceAggregatorInterface';
+
+export class GivPowerBalanceAggregatorMock
```
I would name it
```suggestion
export class GivPowerBalanceAggregatorAdapterMock
```

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorAdapter.ts:
```diff
@@ -0,0 +1,108 @@
+    address: balance.address,
+
+    // TODO convert wei to eth
+    balance: Number(balance.balance),
```
The number cannot hold the balance in wei, it doesn't have enough precision!
We can do the same as we used to do on fetching balance from the subgraph directly.

mohammadranjbarz replied:
Done

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorAdapter.ts:
```diff
@@ -0,0 +1,108 @@
+  ): Promise<GetBalanceOfAddressesResponse> {
+    try {
+      const response = await axios.post(
+        `${this.baseUrl}/givpower-balance-aggregator/power-balance/by-timestamp`,
```
This `/givpower-balance-aggregator` part is not mandatory. For our staging case, we must append it to the URL set in the env

mohammadranjbarz replied:
Done

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+export interface GetBalanceOfAddressesInputParams
+  extends GetLeastIndexedBlockTimeStampInputParams {
+  timestamp: number; // sample: 1691739112
+  addresses: string; // comma separated sample: walletAddress1,walletAddress2,..
```
Let's transfer it to appropriate format in the adapter
```suggestion
  addresses: string[]; 
```

mohammadranjbarz replied:
Done

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorAdapter.ts:
```diff
@@ -0,0 +1,108 @@
+      const response = await axios.post(
+        `${this.baseUrl}/givpower-balance-aggregator/power-balance/updated-after-date`,
+        params,
+      );
```
Pagination is missed! Also, there are some considerations which must be observed as mentioned here
https://github.com/Giveth/impact-graph/issues/1094#issue-1844862318

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorAdapter.ts:
```diff
@@ -0,0 +1,108 @@
+        `${this.baseUrl}/givpower-balance-aggregator/fetch-state/least-indexed-block-timestamp`,
+        params,
+      );
+      return response.data;
```
Are you sure it's number and not string?

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+export interface GetLeastIndexedBlockTimeStampInputParams {
+  networks?: string; // comma separated sample: 100,420
+  network?: string | number; // comma separated sample: 100
```
It's not comma separated 
I would say let's skip `network` and just use `networks`

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+export interface GetLeastIndexedBlockTimeStampInputParams {
```
```suggestion
export interface BANetworksParams {
```

aminlatifi replied:
I would add BA prefix to all params
Also for this one, the name must be more general as other parameter types are extending it

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+  network?: string | number; // comma separated sample: 100
+}
+
+export interface GetLatestBalanceOfAnAddressInputParams
```
```suggestion
export interface BALatestAddressBalanceParams
```

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+  addresses: string; // comma separated sample: walletAddress1,walletAddress2,..
+}
+
+export interface GetBalanceOfAnAddressesResponse {
```
```suggestion
export interface IBAAddressBalanceResponse {
```

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+  updatedAt: Date; // sample: "2023-08-10T16:18:02.655Z"
+  networks: number[]; // sample: [100]
+}
+export type GetBalanceOfAddressesResponse = GetBalanceOfAnAddressesResponse[];
```
```suggestion
export type BAAddressesBalanceResponse = BAAddressBalanceResponse[];
```

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+}
+export type GetBalanceOfAddressesResponse = GetBalanceOfAnAddressesResponse[];
+export type GetBalancesUpdatedAfterASpecificDateResponse =
+  GetBalanceOfAddressesResponse;
```
Is it needed to define new type?

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+export type GetBalancesUpdatedAfterASpecificDateResponse =
+  GetBalanceOfAddressesResponse;
+
+export interface GetBalancesUpdatedAfterASpecificDateInputParams
```
```suggestion
export interface BABalancesUpdatedAfterDateParams
```

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+    params: GetBalanceOfAddressesInputParams,
+  ): Promise<GetBalanceOfAddressesResponse>;
+
+  getLatestBalanceOfAnAddress(
```
```suggestion
  getLatestBalanceSingle(
```

mohammadranjbarz replied:
Done

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+}
+
+export interface GivPowerBalanceAggregatorInterface {
+  getBalanceOfAddresses(
```
```suggestion
  getAddressesBalance(
```

mohammadranjbarz replied:
Done

aminlatifi commented on src/adapters/givPowerBalanceAggregator/givPowerBalanceAggregatorInterface.ts:
```diff
@@ -0,0 +1,48 @@
+    params: GetLatestBalanceOfAnAddressInputParams,
+  ): Promise<GetBalanceOfAnAddressesResponse>;
+
+  getBalancesUpdatedAfterASpecificDate(
```
```suggestion
  getBalancesUpdatedAfterDate(
```

mohammadranjbarz replied:
Done

aminlatifi commented on src/services/cronJobs/fillSnapshotBalances.ts:
```diff
@@ -46,88 +49,122 @@ export async function addFillPowerSnapshotBalanceJobsToQueue() {
   while (!isFinished) {
     const powerBoostings = await getPowerBoostingSnapshotWithoutBalance(
-      100,
+      20,
```
Why have we reduced it? it's for fetching values from our internal db.
```suggestion
      100,
```

mohammadranjbarz replied:
Done

aminlatifi commented on src/services/cronJobs/fillSnapshotBalances.ts:
```diff
@@ -46,88 +49,122 @@ export async function addFillPowerSnapshotBalanceJobsToQueue() {
+      logger.debug('job data', jobData);
+      fillSnapshotBalanceQueue.add(jobData);
+    });
+  }
```
This logic must be outside of the `while` loop. As I see we have reordered some logics in this file which may cause bugs

aminlatifi replied:
One thing that looks obvious is that the `groupByTimestamp` is never emptied... and in each iteration of loop, all its content will be processed that part of it is repetitive.

mohammadranjbarz replied:
Now we have 
```
  const groupByTimestamp: Record<
    number,
    {
      userId: number;
      powerSnapshotId: number;
      walletAddress: string;
    }[]
  > = {};
  ```
  In the beginning of function, so we don't need to make  `groupByTimestamp` empty

aminlatifi commented on src/services/cronJobs/fillSnapshotBalances.ts:
```diff
@@ -46,88 +49,122 @@ export async function addFillPowerSnapshotBalanceJobsToQueue() {
+              )!.userId,
             };
           }),
         );
```
What if the balance of some users are not returned!

mohammadranjbarz replied:
It's not important we find through our addresses from DB not balances , and we are sure we find a user because we just get balances of addresses that we have sent

aminlatifi replied:
If the balance aggregator doesn't have the address of a user, there will be no item in the response corresponding to that user's address. In this case, the user balance in the snapshot will remain `null` and we will query it again!

aminlatifi replied:
It must have the balance! otherwise there would be an issue... let's keep it as it is

aminlatifi commented on src/services/cronJobs/fillSnapshotBalances.ts:
```diff
@@ -46,88 +49,122 @@ export async function addFillPowerSnapshotBalanceJobsToQueue() {
       walletAddress: string;
     }[];
   } = {};
+
```
We must fetch the least indexed block timestamp here and store it in a variable. Then compare the snapshot time to this value and fetch snapshot balance if least indexed timestamp is greater or equal to snapshot time

mohammadranjbarz replied:
Done

CarlosQ96 committed:
add donation currency stats query

CarlosQ96 committed:
fix the latestprojectUpdate query

mohammadranjbarz committed:
WIP Make some refactors and changing function names

related to #1094

mohammadranjbarz committed:
WIP Make some refactors and changing function names

related to #1094

mohammadranjbarz committed:
WIP Make some refactors and changing function names

related to #1094

mohammadranjbarz committed:
WIP Make some refactors and changing function names

related to #1094

mohammadranjbarz committed:
Make tests work

related to #1094

mohammadranjbarz committed:
Make tests work

related to #1094

aminlatifi committed:
Added first draft for the getting instant power balance

aminlatifi committed:
Added first draft for the getting instant power balance

mohammadranjbarz committed:
Merge pull request #1100 from Giveth/feat/instant_balance_aggregator_support

Added first draft for the getting instant power balance

mohammadranjbarz committed:
Don't try to update powerSnapshotBalances if balance aggregator is not synced for a while

related to #1094

mohammadranjbarz committed:
Don't try to update powerSnapshotBalances if balance aggregator is not synced for a while

related to #1094

mohammadranjbarz committed:
Refactor fillPowerSnapshotBalance

related to #1094

mohammadranjbarz committed:
Refactor fillPowerSnapshotBalance

related to #1094

aminlatifi committed:
Updated instantBoostingService test

aminlatifi committed:
Updated instantBoostingService test

aminlatifi committed:
Merge remote-tracking branch 'origin/f_1094_integrate_with_balance_aggregator_microservice' into f_1094_integrate_with_balance_aggregator_microservice

mohammadranjbarz committed:
Add prefix of balance aggregator to BALANCE_AGGREGATOR_BASE_URL in config

related to #1094

mohammadranjbarz committed:
Add prefix of balance aggregator to BALANCE_AGGREGATOR_BASE_URL in config

related to #1094

mohammadranjbarz committed:
Merge branch 'f_1094_integrate_with_balance_aggregator_microservice' of github.com:Giveth/impact-graph into f_1094_integrate_with_balance_aggregator_microservice

mohammadranjbarz committed:
https://github.com/Giveth/impact-graph/pull/1095#discussion_r1299400780

related to #1094

mohammadranjbarz committed:
https://github.com/Giveth/impact-graph/pull/1095#discussion_r1299400780

related to #1094

aminlatifi committed:
Added new migrations to drop and recreate instant power related tables

aminlatifi committed:
Added new migrations to drop and recreate instant power related tables

CarlosQ96 committed:
Merge pull request #1099 from Giveth/feature_donors_count_analytics

Add donation currency stats query

CarlosQ96 committed:
further optimize the query to not do function twice

CarlosQ96 committed:
remove one group by

mohammadranjbarz committed:
Fix isWalletAddressInPurpleList webservice to just return addresses of verified active projects, and not case-sensitive

related to https://github.com/Giveth/giveth-dapps-v2/issues/2861#issuecomment-1682321196

aminlatifi committed:
Rename and refactor

aminlatifi committed:
Rename and refactor

CarlosQ96 committed:
Merge pull request #1101 from Giveth/f_2861_fix_isWalletAddressInPurpleList_endpoint

Fix isWalletAddressInPurpleList webservice to just return addresses of verified active projects, and not case-sensitive

CarlosQ96 committed:
improve query sorting

CarlosQ96 committed:
Merge pull request #1088 from Giveth/feature_optimize_project_update_queries

Optimize latest project update query

aminlatifi committed:
Fixed import bug had remained after refactoring

aminlatifi committed:
Fixed import bug had remained after refactoring

aminlatifi committed:
Merge pull request #1070 from Giveth/dependabot/npm_and_yarn/word-wrap-1.2.4

Bump word-wrap from 1.2.3 to 1.2.4

aminlatifi committed:
Merge pull request #1031 from Giveth/dependabot/npm_and_yarn/apollo/server-4.7.4

Bump @apollo/server from 4.3.1 to 4.7.4

aminlatifi committed:
Merge pull request #993 from Giveth/dependabot/npm_and_yarn/vm2-3.9.19

Bump vm2 from 3.9.17 to 3.9.19

aminlatifi committed:
Merge pull request #1052 from Giveth/dependabot/npm_and_yarn/semver-5.7.2

Bump semver from 5.7.1 to 5.7.2

aminlatifi committed:
Updated redis docker version on production and staging env

aminlatifi committed:
Merge remote-tracking branch 'origin/f_1094_integrate_with_balance_aggregator_microservice' into f_1094_integrate_with_balance_aggregator_microservice

aminlatifi committed:
Fixed some issues on fetching instant balances

mohammadranjbarz committed:
Change logic of filling balances of snapshots

related to #1094

aminlatifi committed:
Fixed issue in time comparison on filling power balance snapshots

aminlatifi committed:
Increase test sleep to avoid failure

aminlatifi committed:
Handle snapshots in same second
Added jobId to fill-snapshot-balance-aggregator tasks
